lib/config/dynamic/admiral/source.go

```diff
@@ -0,0 +1,428 @@
+	}
+}
+
+var DefaultDiscovery = &productDiscovery{}
+
+type source struct {
+	mu      sync.Mutex
+	d       discovery
+	sess    *session.Session
+	vchID   string
+	lastCfg *vchcfg.VirtualContainerHostConfigSpec
+
+	// cached values to minimize running discovery
+	projs []string
+	c     *client.Admiral
+}
+
+// Get returns the dynamic config portion from an Admiral instance. For now,
+// this is empty pending details from the Admiral team.
+func (a *source) Get(ctx context.Context) (*vchcfg.VirtualContainerHostConfigSpec, error) {
+	c, projs, err := a.discover(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	wl, err := a.whitelist(ctx, c, projs)
+	if err != nil {
+		return nil, err
+	}
+
+	newCfg := &vchcfg.VirtualContainerHostConfigSpec{
+		Registry: vchcfg.Registry{RegistryWhitelist: wl},
+	}
+
+	if !a.diff(newCfg) {
+		err = dynamic.ErrConfigNotModified
+	}
+
+	a.lastCfg = newCfg
+	return newCfg, err
+}
+
+func (a *source) diff(newCfg *vchcfg.VirtualContainerHostConfigSpec) bool {
+	if a.lastCfg == nil {
+		if newCfg == nil {
+			return false
+		}
+
+		return true
+	}
+
+	if newCfg == nil {
+		return true
+	}
+
+	// compare whitelists
+	if len(a.lastCfg.RegistryWhitelist) != len(newCfg.RegistryWhitelist) {
+		return true
+	}
+
+	for _, w1 := range a.lastCfg.RegistryWhitelist {
+		found := false
+		for _, w2 := range newCfg.RegistryWhitelist {
+			if w2 == w1 {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (a *source) discover(ctx context.Context) (*client.Admiral, []string, error) {
+	a.mu.Lock()
+	c := a.c
+	a.mu.Unlock()
+
+	if c != nil {
+		projs, err := a.projects(ctx, c)
+		if err == nil && len(projs) > 0 {
+			return c, projs, nil
+		}
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	if c != a.c {
+		return a.c, a.projs, nil
+	}
+
+	log.Infof("running product VM discovery")
+	vms, err := a.d.Discover(ctx, a.sess)
+	if err != nil {
+		log.Errorf("error running product VM discovery: %s", err)
+		return nil, nil, err
+	}
+
+	// find the admiral host that has the VCH
+	for _, v := range vms {
+		values, err := keys(ctx, v, []string{admiralEndpointKey, admiralTokenKey})
+		if err != nil {
+			log.Warnf("keys not found in product VM %s: %s", v, err)
+			continue
+		}
+
+		token := values[admiralTokenKey]
+		if token == "" {
+			// not a useable product installation
+			log.Warnf("empty token set in product VM %s", v)
+			continue
+		}
+
+		u, err := url.Parse(values[admiralEndpointKey])
+		if err != nil {
+			log.Warnf("ignoring bad admiral endpoint %s in product VM %s: %s", values[admiralEndpointKey], v, err)
+			continue
+		}
+
+		rt := rtclient.NewWithClient(u.Host, u.Path, []string{u.Scheme}, admClient)
+		rt.DefaultAuthentication = &admiralAuth{token: token}
+		c := client.New(rt, strfmt.Default)
+		projs, err := a.projects(ctx, c)
+		if err == nil && len(projs) > 0 {
+			log.Infof("using Admiral endpoint at %s", u)
+			a.c = c
+			a.projs = projs
+			return c, projs, nil
+		}
```

Comment:
can we log the error before we look for the next admiral - be good to know why we reject it.